src/theory/bags/theory_bags_type_rules.h

```diff
@@ -0,0 +1,303 @@
+    Assert(n.getKind() == kind::UNION_DISJOINT);
+    return NormalForm::checkNormalConstant(n);
+  }
+}; /* struct BinaryOperatorTypeRule */
+
+struct IsIncludedTypeRule
+{
+  static TypeNode computeType(NodeManager* nodeManager, TNode n, bool check)
+  {
+    Assert(n.getKind() == kind::BAG_IS_INCLUDED);
+    TypeNode bagType = n[0].getType(check);
+    if (check)
+    {
+      if (!bagType.isBag())
+      {
+        throw TypeCheckingExceptionPrivate(
+            n, "BAG_IS_INCLUDED operating on non-bag");
+      }
+      TypeNode secondBagType = n[1].getType(check);
+      if (secondBagType != bagType)
+      {
+        if (!bagType.isComparableTo(secondBagType))
+        {
+          throw TypeCheckingExceptionPrivate(
+              n, "BAG_IS_INCLUDED operating on bags of different types");
+        }
+      }
+    }
+    return nodeManager->booleanType();
+  }
+}; /* struct IsIncludedTypeRule */
+
+struct CountTypeRule
+{
+  static TypeNode computeType(NodeManager* nodeManager, TNode n, bool check)
+  {
+    Assert(n.getKind() == kind::BAG_COUNT);
+    TypeNode bagType = n[1].getType(check);
+    if (check)
+    {
+      if (!bagType.isBag())
+      {
+        throw TypeCheckingExceptionPrivate(
+            n, "checking for membership in a non-bag");
+      }
+      TypeNode elementType = n[0].getType(check);
+      // TODO(#226): comments from sets
+      //
+      // T : (Bag Int)
+      // B : (Bag Real)
+      // (= (as T (Bag Real)) B)
+      // (= (bag-count 0.5 B) 1)
+      // ...where (bag-count 0.5 T) is inferred
+
+      if (!elementType.isComparableTo(bagType.getBagElementType()))
+      {
+        std::stringstream ss;
+        ss << "member operating on bags of different types:\n"
+           << "child type:  " << elementType << "\n"
+           << "not subtype: " << bagType.getBagElementType() << "\n"
+           << "in term : " << n;
+        throw TypeCheckingExceptionPrivate(n, ss.str());
+      }
+    }
+    return nodeManager->integerType();
+  }
+}; /* struct CountTypeRule */
+
+struct MkBagTypeRule
+{
+  static TypeNode computeType(NodeManager* nm, TNode n, bool check)
+  {
+    Assert(n.getKind() == kind::MK_BAG);
+    if (check)
+    {
+      if (n.getNumChildren() != 2)
+      {
+        std::stringstream ss;
+        ss << "operands in term " << n << " are " << n.getNumChildren()
+           << ", but MK_BAG expects 2 operands.";
+        throw TypeCheckingExceptionPrivate(n, ss.str());
+      }
+      TypeNode type1 = n[1].getType(check);
+      if (!type1.isInteger())
+      {
+        std::stringstream ss;
+        ss << "MK_BAG expects an integer for " << n[1] << ". Found" << type1;
+        throw TypeCheckingExceptionPrivate(n, ss.str());
+      }
+    }
+
+    return nm->mkBagType(n[0].getType(check));
+  }
+
+  static bool computeIsConst(NodeManager* nodeManager, TNode n)
+  {
+    Assert(n.getKind() == kind::MK_BAG);
+    // for a bag to be a constant, both the element and its multiplicity should
```

Comment:
Multiplicity needs to be positive as well.  Use `n[1].getConst<Rational>().sgn()==1` I believe.

Reply:
I am not sure about this. I see `(mk-bag "A" -5)` is a constant but not in a normal form. 
This node can be rewritten as `emptybag` using the rewriter for bags, which converts it to a normal form. 
What do you suggest?

Reply:
It is not in normal form and thus is not constant. Indeed, the rewriter will rewrite nodes like the above to a constant.  

Notice that `isConst()` returns true if the node is a canonical representation of a constant. Nodes that are equivalent or could be rewritten to a constant are not necessary constants.

Reply:
Ok, I always got confused with this. `isConst` is not the ideal name here. `isNormalForm` better reflects the meaning.

Reply:
Done by adding the sign check to the condition